mapie/quantile_regression.py

```diff
@@ -0,0 +1,425 @@
+from typing import Optional, Tuple, Union, cast, Iterable, List
+
+import numpy as np
+from sklearn.base import RegressorMixin, clone
+from sklearn.linear_model import QuantileRegressor
+from sklearn.pipeline import Pipeline
+from sklearn.utils.validation import (
+    indexable,
+    check_is_fitted,
+    _num_samples,
+    _check_y,
+)
+from typing_extensions import TypedDict
+
+from ._typing import ArrayLike, NDArray
+from .utils import (
+    check_alpha_and_n_samples,
+    check_n_features_in,
+    check_null_weight,
+    fit_estimator,
+    check_lower_upper_bounds,
+)
+from ._compatibility import np_quantile
+from .regression import MapieRegressor
+
+
+class MapieQuantileRegressor(MapieRegressor):
+    """
+    This class implements the conformalized quantile regression strategy
+    to make conformal predictions. The only valid ``method`` is "quantile"
+    and the only valid default ``cv`` is "split".
+    References
+    ----------
+    Yaniv Romano, Evan Patterson and Emmanuel J. Candès.
+    "Conformalized Quantile Regression"
+    https://proceedings.neurips.cc/paper/2019/hash/5103c3584b063c431bd1268e9b5e76fb-Abstract.html
+    """
+    valid_methods_ = ["quantile"]
+    fit_attributes = [
+        "estimators_",
+        "conformity_scores_",
+        "n_samples_calib",
+        "n_features_in_"
+    ]
+
+    Params = TypedDict(
+        "Params",
+        {
+            "loss_name": str,
+            "alpha_name": str
+        },
+    )
+    quantile_estimator_params = {
+        "GradientBoostingRegressor": Params(
+            loss_name="loss",
+            alpha_name="alpha"
+        ),
+        "QuantileRegressor": Params(
+            loss_name="quantile",
+            alpha_name="quantile"
+        ),
+        "HistGradientBoostingRegressor": Params(
+            loss_name="loss",
+            alpha_name="alpha"
+        ),
+        "LGBMRegressor": Params(
+            loss_name="objective",
+            alpha_name="alpha"
+        )
+    }
+
+    def __init__(
+        self,
+        estimator: Optional[Union[RegressorMixin, Pipeline]] = None,
+        method: str = "quantile",
+        alpha: float = 0.1,
+        cv: Optional[Union[int, str]] = None,
+    ) -> None:
+        super().__init__(
+            estimator=estimator,
+            method=method,
+            cv=cv
+        )
+        self.alpha = alpha
+
+    def _check_alpha(
+        self,
+        alpha: float,
+    ) -> NDArray:
+        """
+        Perform several checks on the alpha value and changes it from
+        a float to an ArrayLike
+
+        Parameters
+        ----------
+        alpha : float
+            Can only be a float value between 0 and 0.5.
+            Represent the uncertainty of the confidence interval.
+            Lower alpha produce larger (more conservative) prediction
+            intervals. Alpha is the complement of the target coverage level.
+            Only used at prediction time. By default 0.1
```

Comment:
To remove as `MapieQuantileRegressor` also uses it during fit.

Reply:
done